### PR TITLE
Fix gradient accumulation under FSDP2: defer cross-replica all-reduce

### LINKED ISSUE #161: trainer.grad_accum_iter saves no communication under FSDP2: ddp_sync_grad is a no-op for fully_shard modules
**Summary.** The training loop wraps each micro-step in `distributed.ddp_sync_grad(model, is_boundary)` to defer gradient synchronization to the last micro-step of an accumulation window (`cosmos_framework/trainer/__init__.py`, `training_step`). The context manager only handles `DistributedDataParallel` instances (`cosmos_framework/utils/distributed.py`). With `trainer.distributed_parallelism="fsdp"` the model is a set of `fully_shard` (FSDP2) modules, the isinstance check fails, and the context silently does nothing. Every micro-step then runs the full gradient reduction. Accumulation enlarges the batch but saves no communication.

**Measurement** (2x 8xA100-80GB, Ethernet inter-node, HSDP shard=8 replicate=2, action-policy recipe, batch 64/rank):

| Config | s/iter | samples/s |
|---|---|---|
| grad_accum_iter=1 | 110.2 | 9.3 |
| grad_accum_iter=4 | 430.6 | 9.5 |

430.6 s is four accum=1 steps back to back.

**Fix, validated on hardware** (same instance type, a different rental whose accum=1 step is 22.2 s): defer only the cross-replica all-reduce via `FSDPModule.set_requires_all_reduce(False)` on non-boundary micro-steps. The intra-node reduce-scatter still runs each micro-step, so gradients stay sharded and per-rank memory does not grow. With the fix, accum=4 steps take 54.3 / 54.3 / 54.9 s against 88.8 s (4 x 22.2) for the unamortized path. The two measurements decompose to compute 10.8 s + sync 11.4 s per micro-batch, throughput 1.63x at accum=4. `set_requires_gradient_sync(False)` also works but keeps gradients unsharded across the window, which adds the full unsharded gradient bytes per rank.

PR with the fix follows from `Sullivan07043:fix-fsdp2-accum`.

**Environment.** cosmos-framework main (ee58e414), torch 2.10 cu128, NCCL 2.27.5, 16 ranks over 2 nodes.

### PR DESCRIPTION
**Pull request by Sullivan07043**
Fixes #161 (`ddp_sync_grad` is a silent no-op under FSDP2, so `trainer.grad_accum_iter` saves no communication).

**Change.** Add an FSDP2 branch to `ddp_sync_grad`: on non-boundary micro-steps, defer only the cross-replica all-reduce via `FSDPModule.set_requires_all_reduce(enabled, recurse=False)` on every `FSDPModule`. Restore it in the `finally` block. The intra-node reduce-scatter still runs each micro-step, so gradients stay sharded and per-rank memory does not grow. DDP behavior is unchanged.

**Validation** (2x 8xA100-80GB, Ethernet, HSDP shard=8 replicate=2, batch 64/rank, fleet whose accum=1 step is 22.2 s):

| Config | s/step |
|---|---|
| accum=1 | 22.2 |
| accum=4, without this fix (measured behavior = 4x) | 88.8 |
| accum=4, with this fix | 54.3 / 54.3 / 54.9 |

Decomposition from the two measured points: compute 10.8 s + sync 11.4 s per micro-batch. Throughput 1.63x at accum=4. Loss stayed in the expected band.

One commit, 18 added lines, no API change.

### CODE REVIEW

**Comment by lfengad** [comment]
Thanks for the fix — the overall approach makes sense to me. Just two small things I think we should address before merging:

1. The FSDP branch always restores `set_requires_all_reduce(True)`, instead of restoring the previous state like the DDP branch does. Could we preserve the old state here as well? Otherwise nested or external sync contexts could be changed unexpectedly.
2. The FSDP modules are updated before entering the `try/finally`. If one of those calls raises midway, the modules already updated would not be restored. Could we move that setup into the protected block?

Thanks!

**Comment by Sullivan07043** [comment]
Thanks for the review — both points addressed in 799dbba:

1. The context now captures each FSDP module's prior `all_reduce_grads` (mirroring the setter's write target) and restores exactly that value on exit, so nested or external sync contexts keep their state. The DDP branch behavior is unchanged; it now restores through a mutation flag for the same exactness.
2. All module mutations happen inside the `try`/`finally` block, recording `(module, prev)` as they land — if the setup loop raises midway, the `finally` restores exactly the modules already changed.

Verified with a single-rank `fully_shard` semantic test: flip/restore, nested contexts preserving an outer `False`, exception-path restore, and `enabled=True` as a state no-op.

File: cosmos_framework/utils/distributed.py [hunk 1]
```diff
@@ -241,19 +241,50 @@ def ddp_sync_grad(model, enabled):
...
+        else:
+            # FSDP2 (fully_shard) branch. Without this, gradient accumulation
+            # reduces gradients on EVERY micro-step and saves no communication
+            # (measured: accum=4 step = 4x the accum=1 step on Ethernet HSDP).
+            # We defer only the cross-replica all-reduce to the boundary
+            # micro-step. The intra-node reduce-scatter still runs every
+            # micro-step, so gradients stay sharded and VRAM does not grow
+            # (set_requires_gradient_sync(False) would keep gradients
+            # unsharded: +stored-grad-bytes per rank).
+            from torch.distributed.fsdp import FSDPModule
```
Comment (lfengad) [inline cosmos_framework/utils/distributed.py:272]: Small thing: could we guard this import with `dist.is_available()`? The rest of this module does that for its `torch.distributed` imports. Every non-DDP model now takes this branch, including plain single-GPU runs, so a build without distributed would hit an ImportError on every micro-step.

An early `if not dist.is_available() or not dist.is_initialized(): yield; return` at the top of the branch would handle it, and it also skips the per-step `model.modules()` walk when there's nothing to do. Thanks!

Comment (Sullivan07043) [reply cosmos_framework/utils/distributed.py:272]: Good catch, thanks - added in b62c79b. The branch now starts with `if not dist.is_available() or not dist.is_initialized(): yield; return`, so a build without distributed never reaches the fsdp import, and single-process runs skip the per-step `modules()` walk. Verified with dist uninitialized: the context yields without walking `modules()`, and exceptions propagate cleanly through the early exit. Also merged main to bring the branch up to date.

**Review by lfengad — APPROVED** [review]

**Review by lfengad — CHANGES_REQUESTED** [review]
Wait for the latest sync release of the framework code, then could merge. Thx!

**Review by foreverlms — APPROVED** [review]
LGTM

**Review by lfengad — APPROVED** [review]
